<a href="https://colab.research.google.com/github/majorquev/Introduccion_a_ciencia_de_datos_R/blob/master/Ayudant%C3%ADas/Ayudant%C3%ADa1/A1_02_listas_y_dataframes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Ayudantía 1**

**Ayudante:** Bastián Aballay Leiva<br>
**E-mail ayudante:** bastian.aballayleiva@gmail.com



# **2 Listas y Dataframes**

## **2.1 Listas**

Hasta aquí contamos con dos tipos de objetos para almacenar datos; vectores y matrices. Sin embargo, en ambos casos se tiene la limitante de que no es posible combinar distintos tipos de datos en diferentes entradas. Un objeto que permite una mayor flexibilidad son las listas. En R, una lista es básicamente un arreglo unidimensional, donde en cada entrada es posible almacenar distintos **objetos** (los cuales a su vez pueden almacenar distintos tipos de datos (u otros objetos).
En resumen,

   * ¿Qué son?<br>
    Arreglos unidimensionales que guardan distintos tipos de datos u objetos.<br>

   * ¿Cómo se declaran?<br>
    Se pueden declarar mediante el comando list() o combinando listas existentes. <br>
    
   * ¿Cómo se accede a sus elementos?<br>
    Usando el **operador [ ]** (retorna lista), o bien usando **[ [ ] ]** (retorna el objeto) - "squarebrackets". 

    En cada caso se puede indicar la posición o nombre del objeto que se quiere rescatar.<br>
    
    Otra manera es mediante el símbolo $ seguido del nombre del objeto.



In [3]:
# 2. Listas y dataframes ------------
# 2.1 Listas  ------------
# Cómo declarar una lista (comando list() )
# num, letra, pl, mat, lista

lista1 <- list(
  num = 6,
  letra = "A",
  vector = c(10,20,30),
  matriz = diag(8,4),
  lista = list(a = "hola", b = 18)
)

In [4]:
?str

In [5]:
# estructura
str(lista1)

List of 5
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18


In [9]:
# Creamos una matriz diagonal (Algebra lineal!)
diag(8,4)

[,1] [,2] [,3] [,4]
[1,] 1    0    0    0   
[2,] 0    1    0    0   
[3,] 0    0    1    0   
[4,] 0    0    0    1

In [6]:
# Obtenemos la clase del objeto
class(lista1)

[1] "list"

In [7]:
# para saber el largo de la lista/vector
length(lista1)

[1] 5

In [10]:
# Para obtener los nombres de la lista
names(lista1)

[1] "num"    "letra"  "vector" "matriz" "lista"

In [12]:
# Ejemplo de una lista simple
lista_test <- list("Hola",42,TRUE)
lista_test

names(lista_test)

[[1]]
[1] "Hola"

[[2]]
[1] 42

[[3]]
[1] TRUE

NULL

In [14]:
str(lista1)

List of 5
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18


In [19]:
# Podemos combinar listas
lista2 <- list(
  primos = c(2,3,5,7),
  amigos = c(220,284)
)

# Print lista (imprime la lista de manera nativa)
lista2

# Estructura de lista
str(lista2)

# Print Lista con método print
print(lista2)


$primos
[1] 2 3 5 7

$amigos
[1] 220 284

List of 2
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284
$primos
[1] 2 3 5 7

$amigos
[1] 220 284



In [17]:
lista3 <- c(lista1,lista2)

lista3

$num
[1] 6

$letra
[1] "A"

$vector
[1] 10 20 30

$matriz
     [,1] [,2] [,3] [,4]
[1,]    8    0    0    0
[2,]    0    8    0    0
[3,]    0    0    8    0
[4,]    0    0    0    8

$lista
$lista$a
[1] "hola"

$lista$b
[1] 18


$primos
[1] 2 3 5 7

$amigos
[1] 220 284

### **Acceso a los elementos de una lista**

Al igual que un vector, una lista es un arreglo unidimensional, por lo que es esperable que herede la misma forma de acceso.

#### **Acceso mediante posición**

En R, para acceder al "slot" de una lista, podemos utilizar `[]` indicando la posición a la que queremos acceder. Notar que en R distingue el uso de `[]` y `[[]]`. Para ilustrar este comportamiento, consideremosla lista "lista3", donde se observa que `lista3[2]` retorna una lista, mientras que `lista3[[2]]` retorna el character "A"








In [21]:
str(lista3)

List of 7
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284


In [22]:
# Acceso mediante posición
lista3[2] # Retorna una lista con el slot de posición 2.

class(lista3[2])

$letra
[1] "A"

[1] "list"

In [23]:
# Por otro lado
lista3[[2]] # Retorna el character "A"

class(lista3[[2]])

[1] "A"

[1] "character"

In [26]:
lista3[4]
class(lista3[4])

$matriz
     [,1] [,2] [,3] [,4]
[1,]    8    0    0    0
[2,]    0    8    0    0
[3,]    0    0    8    0
[4,]    0    0    0    8

[1] "list"

In [27]:
# Podemos combinar lo que hemos aprendido, por ejemplo, en el 4to slot hay una matriz y queremos obtener la 3ra fila #:= 0 0 8 0
lista3[[4]]
class(lista3[[4]])

[,1] [,2] [,3] [,4]
[1,] 8    0    0    0   
[2,] 0    8    0    0   
[3,] 0    0    8    0   
[4,] 0    0    0    8

[1] "matrix" "array"

In [28]:
lista3[[4]][3,] # 0 0 8 0

[1] 0 0 8 0

#### **Acceso mediante nombre**

Cuando una lista posee nombres, podemos acceder a sus elementos referenciando su nombre, distinguiendo, al igual que por posición, el uso de `[]` y `[[]]`


In [29]:
str(lista3)

List of 7
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284


In [30]:
lista3["matriz"] # Retorna una lista con la matriz
class(lista3["matriz"])

$matriz
     [,1] [,2] [,3] [,4]
[1,]    8    0    0    0
[2,]    0    8    0    0
[3,]    0    0    8    0
[4,]    0    0    0    8

[1] "list"

In [31]:
lista3[["matriz"]] # Retorna la matriz
class(lista3[["matriz"]])


[,1] [,2] [,3] [,4]
[1,] 8    0    0    0   
[2,] 0    8    0    0   
[3,] 0    0    8    0   
[4,] 0    0    0    8

[1] "matrix" "array"

#### **Acceso mediante "$"**

Finalmente, una de las opciones más simples para acceder a los elementos de una lista cuando sus entradas poseen nombre, es referenciando estos objetos anteponiendo el signo "$".Por ejemplo, para acceder al 4to slot, con la matriz, podemos escribir `lista3$matriz`, lo cual nos retorna directamente la matriz (no una lista con la matriz).

In [32]:
lista3$letra

[1] "A"

In [33]:
class(lista3$letra)

[1] "character"

In [34]:
lista3$matriz # Retorna la matriz directamente

[,1] [,2] [,3] [,4]
[1,] 8    0    0    0   
[2,] 0    8    0    0   
[3,] 0    0    8    0   
[4,] 0    0    0    8

In [35]:
# Extraer hola
lista3$lista$a
# OJO!... podría ocurrir esto : abc$asdasdA$asdasda$asdasd$asdasd$asdasdasd$aasdasdad$adasda$hola <- evaluar si "vale la pena"

[1] "hola"

### **Creación de nuevo slots**

Podemos crear nuevos slots de manera sencilla, indicano el nuevo nombre (utilizando `[]` o `$`) y luego asignando el objeto a almacenar en la nueva entrada.

In [36]:
str(lista3)

List of 7
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284


In [37]:
# Creamos un slot llamado "nuevo" utilizadno [[]]
# lista[["<nombre_nuevo_objeto>"]] <- objeto
lista3[["nuevo"]] <- c("vector","de","palabras")
str(lista3)

List of 8
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284
 $ nuevo : chr [1:3] "vector" "de" "palabras"


In [42]:
# Creamos un slot llamdo "otro" utilizando $
lista3$otro <- "Esto es sólo un texto"
str(lista3)


List of 9
 $ num   : num 6
 $ letra : chr "A"
 $ vector: num [1:3] 10 20 30
 $ matriz: num [1:4, 1:4] 8 0 0 0 0 8 0 0 0 0 ...
 $ lista :List of 2
  ..$ a: chr "hola"
  ..$ b: num 18
 $ primos: num [1:4] 2 3 5 7
 $ amigos: num [1:2] 220 284
 $ nuevo : chr [1:3] "vector" "de" "palabras"
 $ otro  : chr "Esto es sólo un texto"


In [41]:
# (Pregunta César: ¿qué sucede con las comillas?)
a <- "hola"
b <- 'hola'
c <- "Hello I'm Bastian"
a == b

c

[1] TRUE

[1] "Hello I'm Bastian"

Por qué utilizar listas:
- Para almacenar contenidos de API
- Para guardar objetos de interés
  - Modelos
  - Gráficos
  - Listas
  - eeeeeeeeeetcccccccccccc.
  

## **2.2 Dataframes**

Un dataframe es un caso particular de una lista, donde los objetos de cada entrada corresponden a vectores de datos del mismo largo. En este caso cada vector corresponde a una columna del data.frame, mientras que la cantidad de filas corresponde al largo común de los vectores de cada entrada. El objeto resultante es una *tabla*, de ahora en adelante *data.frame*, cuyas columnas poseen distintos tipos de datos entre sí. 

 * ¿Cómo se declara un dataframe?<br>
    Mediante la función **data.frame()**, indicando los vectores que conformarán las columnas.
 
 * ¿Cómo se accede a los datos de un data.frame?<br>
    Al ser una lista, se puede acceder de la misma manera. Sin embargo también es posible acceder del modo en que se accede a los elementos de una matriz.

In [46]:
# 5.2 Dataframes ----------
# Ejemplo sencillo de definición de un data.frame
# recordar que 1:3 es lo mismo que c(1,2,3)

df <- data.frame(
  id = 1:3,
  nombre = c("pedro","juan","diego"),
  edad = c(32,33,28)
)

df


id nombre edad
1 1  pedro  32  
2 2  juan   33  
3 3  diego  28

In [48]:
# Acceso a columnas de un data.frame
## Igual que en una lista
df$nombre
df[["nombre"]] #retorna el vector con los nombres, mientras que df["nombres"] retorna un data.frame con los nombres
df[[2]] # ID

[1] "pedro" "juan"  "diego"

[1] "pedro" "juan"  "diego"

[1] "pedro" "juan"  "diego"

In [49]:
df["nombre"] # Nos entrega un data.frame

nombre
1 pedro 
2 juan  
3 diego

In [51]:
##  Ganamos acceso como matriz
# segunda fila de df
df[2,]

# segunda columna y segunda fila
df[2,2]

# Por nombre de columna
df[2,]$nombre

id nombre edad
2 2  juan   33

[1] "juan"

[1] "juan"

In [52]:
#Podemos crear nuevas columnas , del mismmo modo que creamos "slots" en una lista
df$hijos <- c(0,1,4)

df

id nombre edad hijos
1 1  pedro  32   0    
2 2  juan   33   1    
3 3  diego  28   4

In [53]:
# podemos seleccionar más de una columna a la vez. Ej: seleccionar columnas 1 y 3
# Referenciando el nombre de las columnas
df[,c("id","edad")]

id edad
1 1  32  
2 2  33  
3 3  28

In [54]:
# Referenciando las posiciones de las columnas
df[,c(1,3)]

id edad
1 1  32  
2 2  33  
3 3  28

In [56]:
df[,2:3] 
df[,c(2,3)]

nombre edad
1 pedro  32  
2 juan   33  
3 diego  28

nombre edad
1 pedro  32  
2 juan   33  
3 diego  28



---
# **Taller 2**
La siguiente línea de código permite cargar en memoria, un objeto de extensión *.rds que está almacenado en el disco de la máquina en la ruta actual de trabajo.

Este archivo contiene la información de los contagios por COVID-19 en nuestsro país, según las cifras oficiales entregadas por el ministerio de salud actualizadas al 08 de mayo. En el archivo encontrará, **para cada mes, el total de casos diarios** desglosados según:
   * 'casos_nuevos_con_sintomas': casos nuevos con síntomas del día.
   * 'casos_totales' : casos acumulados a la fecha.
   * 'casos_recuperados': casos recuperados acumulados a la fecha.
   * 'fallecidos': casos fallecidos acumulados.
   * 'casos_activos': casos activos a la fecha.
   * 'casos_nuevos_sin_sintomas': casos asintomáticos del día.
   * 'casos_nuevos_totales': suma de  casos asintomáticos más casos con síntomas del día.
 
 Si bien no será utilizados en este trabajo, también econtrán nuevos registros de acuerdo a los últimos cambios en la metodología de conteo. Para más info pueden referirse al siguiente link:
 https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto5
 
   * 'casos_activos_por_fd'
   * 'casos_activos_por_fis'
   * 'casos_recuperados_por_fis'
   * 'casos_recuperados_por_fd' 
   * 'casos_confirmados_recuperados'
   * 'casos_activos_confirmados'
   * 'casos_probables_acumulados'
   * 'casos_activos_probables'




In [57]:
# Carga de casos
casos <- readRDS("covid.rds")

In [60]:
# Qué acabo de importar?
class(casos)

[1] "list"

In [ ]:
str(casos)

## Pregunta 1.1

**1a)** ¿Qué clase de objetos es `casos`? <br>
**1b)** ¿Cuántos elementos posee este objeto?<br>
**1c)** ¿Cuáles son los nombres de este objeto?<br>
**1d)** ¿Qué clase de objeto está almacenado en sl _slot_ correspondiente al mes de junio ("jun") y cuántos elementos posee?<br>
**1e)** Para el mes de junio, ¿a qué clase de objeto corresponde `casos_totales` y cuántos elementos posee?




In [61]:
# P1.1a
## Respuesta 
class(casos)

[1] "list"

In [62]:
# P1.1b
## Respuesta 
length(casos)

[1] 5

In [63]:
# P1.1c
## Respuesta 
names(casos)

[1] "mar" "abr" "may" "jun" "jul"

In [ ]:
# P1.1c
## Respuesta 
casos$jun

In [67]:
length(casos$jun)
# Podemos asumir que dentro de la lista de cada mes existen 15 dataframes asocia
# dos a alguna información específica

[1] 15

In [71]:
# Para el mes de junio, ¿a qué clase de objeto corresponde casos_totales y cuántos elementos posee?
# ALGORITMO 
# 1. Rescatar mes de junio
# casos$jun
# 2. Obtener tabla de casos_totales
# casos$jun$casos_totales
# 3. Contar cantidad de filas, que deben corresponder a la cantidad de días que tenga el mes (con casos_totales)
nrow(casos$jun$casos_totales) 

[1] 30

In [72]:
?nrow

In [73]:
# Y si me interesan las columnas?
ncol(casos$jun$casos_totales)

[1] 2

## Pregunta 1.2

**2a)** ¿Cuántos casos activos existían al 25 de marzo? (indique claramente la expresión de R para obtener el valor solicitado) .<br>
**2b)** Genere cinco nuevas variables llamadas `totales_marzo`,`totales_abril` , `totales_mayo`, `totales_junio` y `totales_julio`, que contengan los dataframes correspondientes al slot `casos_totales` del mes respectivo, (por ejemplo, la variable `totales_julio` debe contener una dataframe de dimensión $8\times2$.<br>
**2c)** Para cada dataframe del paso anterior, genere una nueva columna llamada `mes`, que repita el nombre del mes correspondiente por cada fila.<br>
**2d)** En cada data.frame, agregue una nueva columna llamada `casos_diarios` que contanga la información correspontiende al total de casos nuevos diarios(`casos_nuevos_totales`).


In [ ]:
# P1.2a
## Respuesta 
# 1. Rescatar mes de marzo
# 2. Obtener tabla de casos_activos
# 3. Buscar fila 25 !!!(no funcionará)
# 4. Identificar una metodología bulletproof

In [85]:
casos$mar$casos_activos

total dia_del_mes
1     1   2         
2     1   3         
3     3   4         
4     4   5         
5     5   6         
6     7   7         
7    10   8         
8    15   9         
9    17  10         
10   23  11         
11   33  12         
12   43  13         
13   61  14         
14   75  15         
15  156  16         
16  200  17         
17  235  18         
18  338  19         
19  429  20         
20  531  21         
21  623  22         
22  734  23         
23  903  24         
24 1117  25         
25 1269  26         
26 1562  27         
27 1842  28         
28 2057  29         
29 2285  30         
30 2525  31

In [80]:
# Esto NO sirve, porque la fila 25 corresponde al día 26 del mes.
casos$mar$casos_activos[25,]

total dia_del_mes
25 1269  26

In [82]:
# Verificamos que se cumpla la condición requerida
casos$mar$casos_activos$dia_del_mes == 25

[1] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE
[25] FALSE FALSE FALSE FALSE FALSE FALSE

In [84]:
# El elemento anterior nos permite filtrar el dataframe, entramos por su valor!!
# df[row!!!, col]
casos$mar$casos_activos[casos$mar$casos_activos$dia_del_mes == 25,]

total dia_del_mes
24 1117  25

Genere cinco nuevas variables llamadas totales_marzo,totales_abril , totales_mayo, totales_junio y totales_julio, que contengan los dataframes correspondientes al slot casos_totales del mes respectivo, (por ejemplo, la variable totales_julio debe contener una dataframe de dimensión 8×2.

In [92]:
# P1.2b
totales_marzo <- casos$mar$casos_totales
totales_abril <- casos$abr$casos_totales
totales_mayo <- casos$may$casos_totales
totales_junio <- casos$jun$casos_totales
totales_julio <- casos$jul$casos_totales

In [94]:
totales_julio

total  dia_del_mes
1 282043 1          
2 284541 2          
3 288089 3          
4 291847 4          
5 295532 5          
6 298557 6          
7 301019 7          
8 303083 8

In [102]:
# Para cada dataframe del paso anterior, genere una nueva columna llamada mes, que repita el nombre del mes correspondiente por cada fila.
totales_marzo$mes <- "marzo" # totales_marzo[["mes"]]
totales_abril$mes <- "abril"
totales_mayo$mes <- "mayo"
totales_junio$mes <- "junio"
totales_julio$mes <- "julio"

#totales_marzo
head(totales_julio,3)

total  dia_del_mes mes  
1 282043 1           julio
2 284541 2           julio
3 288089 3           julio

[!]: esta acción anterior "recicla" el elemento asignado


In [108]:
# d) En cada data.frame, agregue una nueva columna llamada casos_diarios que 
# contanga la información correspontiende al total de casos nuevos 
# diarios(casos_nuevos_totales).

totales_marzo[["casos_diarios"]] <- casos$mar$casos_nuevos_totales$total
totales_abril[["casos_diarios"]] <- casos$abr$casos_nuevos_totales$total
totales_mayo[["casos_diarios"]] <- casos$may$casos_nuevos_totales$total
totales_junio[["casos_diarios"]] <- casos$jun$casos_nuevos_totales$total
totales_julio[["casos_diarios"]] <- casos$jul$casos_nuevos_totales$total

head(totales_marzo,3)
head(totales_abril,3)
head(totales_mayo,3)
head(totales_junio,3)
head(totales_julio,3)

total dia_del_mes mes   casos_diarios
1 1     2           marzo 1            
2 1     3           marzo 0            
3 3     4           marzo 2

total dia_del_mes mes   casos_diarios
1 3031  1           abril 293          
2 3404  2           abril 373          
3 3737  3           abril 333

total dia_del_mes mes  casos_diarios
1 17008 1           mayo  985         
2 18435 2           mayo 1427         
3 19663 3           mayo 1228

total  dia_del_mes mes   casos_diarios
1 105159 1           junio 5471         
2 108686 2           junio 3527         
3 113628 3           junio 4942

total  dia_del_mes mes   casos_diarios
1 282043 1           julio 2650         
2 284541 2           julio 2498         
3 288089 3           julio 3548

## Pregunta 1.3

**3a)** Con la ayuda de lafunción rbind() , una las filas de las 5 tablas generadas en la pregunta anterior y guarde la tabla resultante en una variable llamada `casos_hist`.<br>
**3b)** En promedio, ¿Cuántos casos hay diariamente desde el 03 de marzo hasta el 08 de julio?<br>
**3c)** En promedio, ¿Cuántos casos hay diariamente **en cada mes** ?.<br>


In [109]:
# P1.3
## Respuesta
casos_hist <- rbind(
  totales_marzo,
  totales_abril,
  totales_mayo,
  totales_junio,
  totales_julio
) 

In [110]:
head(casos_hist, 5)
tail(casos_hist, 5)

total dia_del_mes mes   casos_diarios
1 1     2           marzo 1            
2 1     3           marzo 0            
3 3     4           marzo 2            
4 4     5           marzo 1            
5 5     6           marzo 1

total  dia_del_mes mes   casos_diarios
125 291847 4           julio 3758         
126 295532 5           julio 3685         
127 298557 6           julio 3025         
128 301019 7           julio 2462         
129 303083 8           julio 2064

In [115]:
# P1.3.b
mean(casos_hist[2:129,"casos_diarios"])
# casos_hist$fecha <- as.Date() TAREA! tip: usar format() !

[1] 2122.344

# Sección 2

La siguiente celda de código permitirá cargar en su sesión de trabajo los siguientes vectores:
* `temp_hombre`: Contiene el registro de la temperatura para una muestra de 650 hombres escogidos al azar.
* `temp_mujer`: Contiene el registro de la temperatura para una muestra de 580 mujeres escogidas al azar.


En ambos vectores, **los identificadores (id) de cada persona están contenidos en el atributo `names()`** del vector respectivo.

In [116]:
temp_h <- readRDS("temp_h.rds")
temp_m <- readRDS("temp_m.rds")

In [120]:
head(temp_h, 5)
class(temp_h)

h_377     h_28     h_20    h_491    h_403 
39.32619 36.52849 36.97699 39.26491 36.19782

[1] "numeric"

In [134]:
head(temp_m, 5)

m_326    m_560     m_98    m_228    m_188 
35.96257 36.04833 36.08442 36.41125 36.40215

In [ ]:
hombres <- as.data.frame(temp_h)
hombres
#colnames(hombres) <- c("id","temperatura")


In [132]:
# mujeres <- 

# 18:00 Octubre 2, 2020

##  Pregunta 2.1

**P1)** A partir de los dos vectores cargados `temp_h` y `temp_m` ustred deberá crear un data.frame llamado **`casos`** con tres columnas y un total de 1230 filas. Las columnas deben ser las siguientes; una con el id del registro, otra con el género y otra con la temperatura registrada. A modo de ejemplo, a continuación se muestra como debería quedar el data.frame:


|id | genero | temperatura|
|---|---|---|
| h_162 | hombre | 36.8 |
| h_140 | hombre | 38.7 |
| h_13  | hombre | 36.1 |
| $\vdots$| $\vdots$ | $\vdots$|
| m_16  | mujer | 39.1 |
| m_502 | mujer | 40.1 |
| m_189 | mujer | 36.8 |

In [117]:
# P2.1. ---------------------------------------------------------------------
## Respuesta:
casos <- rbind(temp_h, temp_m)

Warning message in rbind(temp_h, temp_m):
“number of columns of result is not a multiple of vector length (arg 2)”


## Pregunta 2.2

Adicionalmente, se cuenta con la información de síntomas presentados por cada presona registrada. Esta información se encuentra disponible en el archivo `sintomas.rds`.

**2a)** Cargue el archivo `sintomas.rds` en una variable llamada _sintomas_.<br>
**2b)** ¿Cuál es la clase de este objeto?.<br>
**2c)** Note que cada fila undica con un 1 si la persona presentó el respectivo síntoma y 0 si no. ¿Qué sintomas presentó la persona ubicada en el registro 450?.<br>
**2d)** ¿Qué síntomas y qué temperatura presentó la persona con id "h_537"?. Si es le es de utilidad puede utilizar la función `rownames()`, la cual permite obtener los nombres (id) de las filas de una matriz.<br>

In [ ]:
# P2 .2
## Respuestas:

## Nota de utilidad

Ya sea trabajando con un vector, matriz, data.frame o lista, es posible omitir la selección de alguna(s) de sus posiciones (filas/columnas o slots). Esto se puede llevar a cabo de diversas maneras, por ejemplo:

In [ ]:
# Primero creamos un vector, una matriz, un data.frame  y una lista 

vec <- c(v1 = 1, v2 = 2, v3 = 3,v4 = 4)
mat <- cbind(c(1,1),c(2,2),c(3,3),c(4,4) ) ; rownames(mat) <- c("f1","f2"); colnames(mat) <- c("c1","c2","c3","c4")
dat <- as.data.frame(mat)
lis <- list(s1 = c(1,1), s2 = c(2,2), s3 = c(3,3) , s4 = c(4,4))

### Omisión por posición

In [ ]:
#Si queremos omitir el primer y tercer elmento del vector y la lista
vec[-c(1,3)]
lis[-c(1,3)]

# Misma idea si queremos omitir columnas de una matriz (o data.frame)
mat[,-c(1,3)]
dat[,-c(1,3)]

### Omisión por nombres

In [ ]:
# Por ahora tendremos que ubicar las posiciones de los nombres deseados.

pos_vec <- which(names(vec) %in% c("v1","v3"))
pos_lis <- which(names(lis) %in% c("s1","s3"))
pos_mat <- which(colnames(mat) %in% c("c1","c3"))
pos_dat <- which(names(dat) %in% c("c1","c3"))

vec[-pos_vec]
lis[-pos_lis]

# Misma idea si queremos omitir columnas de una matriz (o data.frame)
mat[,-pos_mat]
dat[,-pos_dat]

Más adelante veremos otras alternivas, como por ejemplo, utilizando funciones del package `dplyr`.

## Pregunta 2.3

Con la ayuda de las funciones `rowSums()` y `colSums()`, identifique lo siguiente.

  **3a)** ¿Cuántos registros presentaron exactamente los 4 síntomas?<br>
  **3b)** ¿Cúantas personas presentaron tos? <br>
  **3c)** ¿Cuál es el síntoma más frecuente?<br>
  

In [ ]:
# P3
##Respuestas:


## Pregunta 2.4

**4a)** La función `as.data.frame()` permite coercionar un objeto de otra clase a `data.frame`.
cree una variable llamada `sintomas_df` con la información  de sintomas pero como data.frame.<br>
**4b)**
Se puede observar que los ids solo están disponibles en los nombres de las filas. Sin embargo esútil tener esta información como una columna más de la tabla. Para ello **cree una nueva columna en la tabla sintomas_df** llamada `id` que contenga los ids de cada registro. 

NOTA: El orden de las filas no ha variado por lo que podemos añadir la columna sin mayores precauciones.<br>
**4c)** ¿Cuántas personas **sólo** presentaron tos y dolor de cabeza?<br>
**4d)** Se sabe que una persona tiene fiebre cuando su temperatura es mayor o igual a los 37.2°. **Cree una nueva columna en la tabla sintomas_df** llamada `tiene_fiebre` que contenga `TRUE` cuando tiene fiebre y `FALSE` en caso contrario. 



In [ ]:
# P4
#Respuestas:
